internal/mage/engine.go

```diff
@@ -48,3 +52,165 @@ func (t Engine) Lint(ctx context.Context) error {
+		}).Build(buildkitRepo)
+		for _, os := range oses {
+			// include each helper for each arch too in case there is a
+			// client/server mismatch
+			for _, arch := range arches {
+				helperBin := util.GoBase(c).
+					WithEnvVariable("GOOS", os).
+					WithEnvVariable("GOARCH", arch).
+					Exec(dagger.ContainerExecOpts{
+						Args: []string{"go", "build", "-o", "./bin/" + sdkHelper, "-ldflags", "-s -w", "/app/cmd/sdk-helper"},
+					}).
+					File("./bin/" + sdkHelper)
+				buildkitBase = buildkitBase.WithFS(
+					buildkitBase.FS().WithFile("/usr/bin/"+sdkHelper+"-"+os+"-"+arch, helperBin),
+				)
+			}
+		}
+		platformVariants = append(platformVariants, buildkitBase)
+	}
+
+	return platformVariants
+}
+
+func (t Engine) Release(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	arches := []string{"amd64", "arm64"}
+	oses := []string{"linux", "darwin"}
+
+	imageRef, err := c.Container().Publish(ctx, engineImageRef, dagger.ContainerPublishOpts{
+		PlatformVariants: engineContainer(c, arches, oses),
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Println("Image published:", imageRef)
+
+	return nil
+}
+
+func localEngine(ctx context.Context, c *dagger.Client) (string, error) {
+	tmpfile, err := os.CreateTemp("", "dagger-engine-export")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile.Name())
+
+	_, err = c.Container().Export(ctx, tmpfile.Name(), dagger.ContainerExportOpts{
+		PlatformVariants: engineContainer(c, []string{runtime.GOARCH}, []string{runtime.GOOS}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	containerName := "test-dagger-engine"
+	volumeName := "test-dagger-engine"
+	imageName := "localhost/test-dagger-engine:latest"
+
+	loadCmd := exec.CommandContext(ctx, "docker", "load", "-i", tmpfile.Name())
+	output, err := loadCmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("docker load failed: %w: %s", err, output)
+	}
+	_, imageID, ok := strings.Cut(string(output), "sha256:")
+	if !ok {
+		return "", fmt.Errorf("unexpected output from docker load: %s", output)
+	}
+	imageID = strings.TrimSpace(imageID)
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"tag",
+		imageID,
+		imageName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker tag: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"rm",
+		"-fv",
+		containerName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker rm: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"run",
+		"-d",
+		"--rm",
+		"-v", volumeName+":/var/lib/buildkit",
+		"--name", containerName,
+		"--privileged",
+		imageName,
+		"--debug",
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker run: %w: %s", err, output)
+	}
+
+	return containerName, nil
+}
+
+func (t Engine) StartLocal(ctx context.Context) error {
```

Comment:
So this is going to be the dev process going forward right? Start this, export, and do your thing?

What happens to the rest of the CI code that uses the Go SDK? Do we need to do this as well? #3704

nit: `Start`? `Dev`?

Reply:
> So this is going to be the dev process going forward right? Start this, export, and do your thing?

Yeah this is the baseline. I also included `engine:test` that does this and then runs all the core/integration tests as a convenience, but in general you can use this target and then export DAGGER_HOST and do whatever you want from there.

> What happens to the rest of the CI code that uses the Go SDK? Do we need to do this as well? #3704

Here's what I'm imagining for CI in general, just as a barebones starting place:
1. For our engine tests, we always build the local engine and run core/integration against that
2. For each SDK we:
   1. Have a run that runs outside of dagger-in-dagger (i.e. the same ways users will actually run the SDK)
      * This gives us coverage of their provisioner code and verifies they work against their current hardcoded engine dependency (i.e. the image ref+sha they use by default)
   1. Have a run that runs inside dagger-in-dagger, where dagger has been provisioned using the latest engine built from local code
      * This lets us verify the SDKs against the latest engine without having to actually push it to a registry anywhere and without having to implement the extra `docker-container` engineconn in every SDK (only needed in go).

I want to double check that there's no weird catch 22 cases where we might want to push engine updates that are backwards incompatible w/ SDKs without necessarily having updated the SDKs yet. I think worst case scenario we could make the test runs of SDKs against the latest engine "warning-only" if they fail.

> nit: Start? Dev?

Sure I'll change to `dev`

Reply:
Just to capture discord discussions, the plan will be:

1. Engine tests will run against `main` engine only
2. Each SDK will be tested via dagger-in-dagger against `main` engine only
3. There will be separate test suites that exercise the provisioning code of the SDKs.
   * These can run dagger-in-dagger once we have support for mounting a docker socket in. In the meantime they may just be manually executed to start with.
   * These tests will provision `main` engine which has been pushed to some local registry. This avoids catch 22s of incompatible api changes
4. When we release an SDK we will also run those provisioner tests except against the actual engine dep in the SDKs, not `main`. These tests will block the release if failed.